### PR TITLE
feat: Constraint implements `Ord`

### PR DESCRIPTION
**Pull request by lmondada**

### CODE REVIEW

**Comment by lmondada** [comment]
I have merged in the changes from the previous PR (#43). The git history is a mess, sorry about that -- but the diff is pretty clear.

Note that I have changed the convention around `Assign` predicates: the first argument to the constraint (LHS) is the variable being assigned to. This is both more in line with how assignments are usually written and also make the ordering more natural (as we want the assigned variable to be the most significant value in the lexicographic order).

File: src/new_api_v04/constraint.rs [hunk 1]
```diff
@@ -194,29 +196,30 @@ where
...
                 .collect::<Result<Vec<_>, _>>()
         };
         match &self.predicate {
             Predicate::Assign(ap) => {
-                let lhs = ap.check_assign(data, &args(self.arity() - 1)?);
-                let Some(ConstraintLiteral::Variable(rhs)) = self.args.last() else {
+                let rhs = ap.check_assign(data, &args(self.arity() - 1)?);
+                let Some(ConstraintLiteral::Variable(lhs)) = &self.args.first() else {
                     panic!("Invalid constraint: rhs of AssignPredicate is not a variable");
                 };
```
Comment (aborgna-q) [inline src/new_api_v04/constraint.rs:215]:
```suggestion
                let lhs = self.assigned_variable();
```

File: src/new_api_v04/predicate.rs [hunk 2]
```diff
@@ -19,80 +19,62 @@
...
+    /// Get Predicate arity
+    fn arity(&self) -> usize;
+}
 
 /// A predicate for pattern matching.
-#[derive(Clone, Debug)]
+///
+/// Assign predicates must be of arity N >= 1 and bind the variable passed as its
+/// first argument.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
```
Comment (aborgna-q) [inline src/new_api_v04/predicate.rs:33]: I'd add a comment about `Assign < Filter`, since it's a strong requirement for the constraint ordering.

**Review by aborgna-q — APPROVED** [review]
Nice!

**Comment by lmondada** [comment]
Thanks!

**Pull request merged**